### ISSUE TITLE
Infer type  for local definitions inside let/where block

### ISSUE THREAD

**Issue by scarf005** [idris-lang/Idris2#3016]
- [x] I have read [CONTRIBUTING.md](https://github.com/idris-lang/Idris2/blob/main/CONTRIBUTING.md).
- [x] I have checked that there is no existing PR/issue about my proposal.

## Summary

Allow inferring type for definitions inside let/where block similar to global type inference for top level declarations.

[from discussion in idris server](https://discord.com/channels/827106007712661524/827106088343175180/1129643731813933138)

## Motivation

Some pain point of refactoring functions into small chunks with let/where blocks were having to manually declare types. With help from @Russoul i found out that holes could be also be used without any names `: ?` to explicitly infer types, however it'd be even better if the type inference could be done without explicit annotation.

### Examples

before:
```idris
-- global type inference, can be omitted
getLength : ? -> ?
-- inferred type is getLength : String -> Nat
getLength str = len where 
  -- local type inference, cannot be omitted
  len : ?
  -- inferred type is len : Nat
  len = length str
```

after:
```idris
getLength str = len where 
  len = length str
```

<details><summary>More examples</summary>

```idris
test str = len + totalLen
  where
    xs : ?
    xs = words str
    len : ?
    len = length xs
    totalLen : ?
    totalLen = sum . (map length) $ xs
```
```idris
test str = len + totalLen
  where
    xs = words str
    len = length xs
    totalLen = sum . (map length) $ xs
```

</details> 

### Technical implementation

As inferring types for local declaration is already possible (using explicit `: ?` notation), It may be achievable by implicitly marking local definitions as holes.

## Alternatives considered

Maybe introduce another syntax sugar like `?=` but i don't think this is the right solution.
```idris
getLength str = len where 
  len ?= length str -- same as len : ?; len = length str
```

## Conclusion

For small definitions holding intermediate computation (like `words xs`), being able to infer them would be immensely useful. As global type inference already works, I'd love to see it be done in local definitions too.

**Comment by gallais** [comment]
Related: #1952

See the related issues mentioned there: being lax enough to allow
declaration-free definitions leads to poorer error messages.

**Comment by andrevidela** [comment]
I wonder how much of this can be handled with better error reporting and how much of this needs to change.

For example, assume we have a `where`-clause `f` which does not have a type declaration. We could synthesize a hole for it as `f : ?`. But in addition to that, in any error reporting involving `f`, we could say 

> `f : ?`, resolved as `f : Int`, does not unify with `Int -> Int`

I think that is not possible right now because `?` won't unify with a concrete type in case of error, this makes sense because if we cannot unify, it means we cannot commit to a single term for the hole. In that case, for the purposes of error reporting, we could chose one for display which would give a slightly more helpful message than "?_ won't unify with `Int -> Int`", which is what we have now.

This gets infinitely more tricky because it assumes that we know the type of `f` at use-site: `Int -> Int` this isn't always the case so we end up with two holes and no way to unify them. 

Not sure if this is solvable at all, maybe I'm missing something, seems to be a good candidate to try something out for the upcoming Developer Meeting

**Comment by dunhamsteve** [comment]
If Idris thinks it is a function declaration, I believe it already is synthesizing `?` for the no arg case, `? -> ?` for the one arg case, etc. (if the args are simple variables) This is done by `holeyArg` in `ProcessDef.idr`. 

For
```idris
getLength str = len where 
  len = length str
```
You can see:
```
LOG declare.def:5: Missing type declaration for the alias Main.getLength. Checking first whether it is a misspelling.
LOG declare.def:5: Not a misspelling: go ahead and declare it! [(Tmp:1:11--1:14, "str")] []
...
LOG declare.def:5: Missing type declaration for the alias Main.len. Checking first whether it is a misspelling.
LOG declare.def:5: Not a misspelling: go ahead and declare it! [] []
LOG declare.type:1: Processing Main.len
LOG declare.type:5: Checking type decl: RigW Main.len : ?
```
(I tweaked one log message, the two extra arrays on the error message are the `args` and `nest`.)

The error we get is:
```
Mismatch between: Nat and (str : String) -> ?_.
```
Here we have `Nat` and the actual type of `len` (with the closed over string added). So I think there is an issue with the extra args that locally defined functions get (from closing over locals), which I thought showed up in `nest`.  Perhaps with the explicitly declared `len : ?` something gets processed in the right context.

**Comment by dunhamsteve** [comment]
No, it's `env` that the vars are coming from in this case, my bad. But we do have a similar issue with parameter blocks:
```idris
parameters (n : Nat) (strs : List String)
  len = List.length strs
```

**Comment by dunhamsteve** [comment]
Ok, I'm slow, but `nest` contains the names of the functions that need the extra args.  This seems to do the trick for all the cases (parameters, the cases above, and the more complicated case from discord):
```
diff --git a/src/TTImp/TTImp.idr b/src/TTImp/TTImp.idr
index 6e50dab44..bd5750272 100644
--- a/src/TTImp/TTImp.idr
+++ b/src/TTImp/TTImp.idr
@@ -790,6 +790,7 @@ definedInBlock ns decls =
 
     defName : Namespace -> ImpDecl -> List Name
     defName ns (IClaim _ _ _ _ ty) = [expandNS ns (getName ty)]
+    defName ns (IDef _ nm _) = [expandNS ns nm]
     defName ns (IData _ _ _ (MkImpData _ n _ _ cons))
         = expandNS ns n :: map (expandNS ns) (map getName cons)
     defName ns (IData _ _ _ (MkImpLater _ n _)) = [expandNS ns n]
```
I'll add a test and PR if all of the tests pass.

**Comment by dunhamsteve** [comment]
Also fixes issues #2782, #2592, and #2593

**Comment by dunhamsteve** [comment]
@scarf005 Did PR #3017 resolve this for you?

**Comment by scarf005** [comment]
![image](https://github.com/idris-lang/Idris2/assets/54838975/8cc95bb9-acd7-4c3a-9626-1c5948f594d1)

@dunhamsteve yes, checked that the issue is fixed in 23694c7e5. thanks for #3017!
